arrow-libs/core/arrow-autoclose/src/commonMain/kotlin/arrow/AutoCloseScope.kt

```diff
@@ -63,43 +63,43 @@ import kotlin.coroutines.cancellation.CancellationException
  */
 public inline fun <A> autoCloseScope(block: AutoCloseScope.() -> A): A {
   val scope = DefaultAutoCloseScope()
+  var finalized = false
   return try {
     block(scope)
-      .also { scope.close(null) }
-  } catch (e: CancellationException) {
-    scope.close(e) ?: throw e
   } catch (e: Throwable) {
-    scope.close(e.throwIfFatal()) ?: throw e
+    finalized = true
+    scope.close(e.throwIfFatal())
+    throw e
+  } finally {
+      if (!finalized) {
+        scope.close(null)
+      }
   }
 }
 
 public interface AutoCloseScope {
+  public fun onClose(release: (Throwable?) -> Unit)
+
+  public fun <A : AutoCloseable> install(autoCloseable: A): A =
+    autoCloseable.also { onClose { autoCloseable.close() } }
+
   public fun <A> autoClose(
     acquire: () -> A,
     release: (A, Throwable?) -> Unit
-  ): A
-
-  @ExperimentalStdlibApi
-  public fun <A : AutoCloseable> install(autoCloseable: A): A =
-    autoClose({ autoCloseable }) { a, _ -> a.close() }
+  ): A = acquire().also { a -> onClose { release(a, it) } }
 }
 
 @PublishedApi
 internal class DefaultAutoCloseScope : AutoCloseScope {
   private val finalizers = Atomic(emptyList<(Throwable?) -> Unit>())
 
-  override fun <A> autoClose(acquire: () -> A, release: (A, Throwable?) -> Unit): A =
-    try {
-      acquire().also { a ->
-        finalizers.update { it + { e -> release(a, e) } }
-      }
-    } catch (e: Throwable) {
-      throw e
-    }
+  override fun onClose(release: (Throwable?) -> Unit) {
+    finalizers.update { it + release }
+  }
 
-  fun close(error: Throwable?): Nothing? {
-    return finalizers.get().asReversed().fold(error) { acc, function ->
-      acc.add(runCatching { function.invoke(error) }.exceptionOrNull())
+  fun close(error: Throwable?) {
+    finalizers.value.asReversed().fold(error) { acc, finalizer ->
```

Comment:
Tried to get the code closer to the style of the corresponding resource scope implementation, mainly for the sake of future readers of the code (since in essence the two are very similar)